File: packages/crawler-py/shroodler/urls.py
from __future__ import annotations

from urllib.parse import parse_qsl, urljoin, urlparse, urlunparse


def origin(url: str) -> str:
    p = urlparse(url)
    netloc = p.hostname or ""
    if p.port:
        netloc = f"{netloc}:{p.port}"
    return f"{p.scheme}://{netloc}"


def is_loopback_or_local(url: str) -> bool:
    host = (urlparse(url).hostname or "").lower()
    if host in {"localhost", "127.0.0.1", "::1", "0.0.0.0"}:
        return True
    if host.endswith(".local"):
        return True
    return False


def same_origin(a: str, b: str) -> bool:
    pa, pb = urlparse(a), urlparse(b)
    port_a = pa.port or (443 if pa.scheme == "https" else 80)
    port_b = pb.port or (443 if pb.scheme == "https" else 80)
    return (
        pa.scheme == pb.scheme
        and (pa.hostname or "").lower() == (pb.hostname or "").lower()
        and port_a == port_b
    )


def normalize_url(base: str, href: str) -> str | None:
    if href is None:
        return None
    href = href.strip()
    if not href or href.startswith(("#", "javascript:", "mailto:", "data:", "tel:")):
        return None
    joined = urljoin(base, href)
    parsed = urlparse(joined)
    if parsed.scheme not in {"http", "https"}:
        return None
    # Drop fragment — fragment-only diffs are duplicates.
    parsed = parsed._replace(fragment="")
    path = parsed.path or "/"
    return urlunparse(parsed._replace(path=path))


def canonical_key(url: str) -> str:
    """Identity used for duplicate detection."""
    p = urlparse(url)
    path = p.path or "/"
    if path != "/" and path.endswith("/"):
        path = path[:-1]
    query_items = sorted(parse_qsl(p.query, keep_blank_values=True))
    query = "&".join(f"{k}={v}" for k, v in query_items)
    host = (p.hostname or "").lower()
    port = p.port
    if port and not ((p.scheme == "http" and port == 80) or (p.scheme == "https" and port == 443)):
        netloc = f"{host}:{port}"
    else:
        netloc = host
    return urlunparse((p.scheme, netloc, path, "", query, ""))


def path_of(url: str) -> str:
    return urlparse(url).path or "/"


def query_param_names(url: str) -> list[str]:
    p = urlparse(url)
    names = []
    seen = set()
    for k, _ in parse_qsl(p.query, keep_blank_values=True):
        if k not in seen:
            seen.add(k)
            names.append(k)
    return names

File: packages/crawler-go/internal/urls/urls.go
package urls

import (
	"net"
	"net/url"
	"sort"
	"strings"
)

func Origin(raw string) string {
	u, err := url.Parse(raw)
	if err != nil {
		return raw
	}
	return u.Scheme + "://" + u.Host
}

func IsLocal(raw string) bool {
	u, err := url.Parse(raw)
	if err != nil {
		return false
	}
	host := strings.ToLower(u.Hostname())
	if host == "localhost" || host == "127.0.0.1" || host == "::1" || host == "0.0.0.0" {
		return true
	}
	return strings.HasSuffix(host, ".local")
}

func SameOrigin(a, b string) bool {
	ua, err1 := url.Parse(a)
	ub, err2 := url.Parse(b)
	if err1 != nil || err2 != nil {
		return false
	}
	pa, _ := port(ua)
	pb, _ := port(ub)
	return ua.Scheme == ub.Scheme && strings.EqualFold(ua.Hostname(), ub.Hostname()) && pa == pb
}

func port(u *url.URL) (string, error) {
	if u.Port() != "" {
		return u.Port(), nil
	}
	if u.Scheme == "https" {
		return "443", nil
	}
	return "80", nil
}

func Normalize(base, href string) string {
	href = strings.TrimSpace(href)
	if href == "" || strings.HasPrefix(href, "#") || strings.HasPrefix(strings.ToLower(href), "javascript:") ||
		strings.HasPrefix(strings.ToLower(href), "mailto:") || strings.HasPrefix(strings.ToLower(href), "data:") {
		return ""
	}
	bu, err := url.Parse(base)
	if err != nil {
		return ""
	}
	ref, err := url.Parse(href)
	if err != nil {
		return ""
	}
	out := bu.ResolveReference(ref)
	if out.Scheme != "http" && out.Scheme != "https" {
		return ""
	}
	out.Fragment = ""
	if out.Path == "" {
		out.Path = "/"
	}
	return out.String()
}

func CanonicalKey(raw string) string {
	u, err := url.Parse(raw)
	if err != nil {
		return raw
	}
	path := u.Path
	if path == "" {
		path = "/"
	}
	if path != "/" && strings.HasSuffix(path, "/") {
		path = strings.TrimSuffix(path, "/")
	}
	q := u.Query()
	keys := make([]string, 0, len(q))
	for k := range q {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	var b strings.Builder
	for i, k := range keys {
		vals := q[k]
		sort.Strings(vals)
		for _, v := range vals {
			if b.Len() > 0 {
				b.WriteByte('&')
			}
			b.WriteString(k)
			b.WriteByte('=')
			b.WriteString(v)
			_ = i
		}
	}
	host := strings.ToLower(u.Hostname())
	p := u.Port()
	if p != "" && !((u.Scheme == "http" && p == "80") || (u.Scheme == "https" && p == "443")) {
		host = net.JoinHostPort(host, p)
	}
	out := u.Scheme + "://" + host + path
	if b.Len() > 0 {
		out += "?" + b.String()
	}
	return out
}

func PathOf(raw string) string {
	u, err := url.Parse(raw)
	if err != nil || u.Path == "" {
		return "/"
	}
	return u.Path
}

func QueryNames(raw string) []string {
	u, err := url.Parse(raw)
	if err != nil {
		return nil
	}
	seen := map[string]bool{}
	var names []string
	for k := range u.Query() {
		if !seen[k] {
			seen[k] = true
			names = append(names, k)
		}
	}
	sort.Strings(names)
	return names
}
